### PR TITLE
refactor(CopyToClipboard): switched from React.Component to React.FC

### PR DESCRIPTION
**Pull request by atroynikov**

### CODE REVIEW

File: src/components/CopyToClipboard/CopyToClipboard.tsx [hunk 1]
```diff
@@ -7,75 +7,52 @@ export enum CopyToClipboardStatus {
...
+export function CopyToClipboard({children, text, timeout = 1000, onCopy}: CopyToClipboardProps) {
+    const resolvedText = React.useMemo(() => String(text), [text]);
+    const [status, setStatus] = React.useState(CopyToClipboard.INITIAL_STATUS);
+    const content = React.useMemo(() => children(status), [children, status]);
+    const timerIdRef = React.useRef<number>();
+    const handleCopy = React.useCallback<Required<ReactCopyToClipboard.Props>['onCopy']>(
+        (copyText, result) => {
+            setStatus(result ? CopyToClipboardStatus.Success : CopyToClipboardStatus.Error);
+
+            if (timerIdRef.current) {
```
Comment (atroynikov) [inline src/components/CopyToClipboard/CopyToClipboard.tsx:26]: Change to `timerIdRef.current !== undefined`

File: src/components/CopyToClipboard/CopyToClipboard.tsx [hunk 2]
```diff
@@ -7,75 +7,52 @@ export enum CopyToClipboardStatus {
...
-            <ReactCopyToClipboard text={String(text)} onCopy={this.handleCopy}>
-                {content}
-            </ReactCopyToClipboard>
-        );
-    }
-
-    componentWillUnmount() {
-        clearTimeout(this.timerId);
+export function CopyToClipboard({children, text, timeout = 1000, onCopy}: CopyToClipboardProps) {
+    const resolvedText = React.useMemo(() => String(text), [text]);
```
Comment (ogonkov) [inline src/components/CopyToClipboard/CopyToClipboard.tsx:18]: useMemo looks redundant

File: src/components/CopyToClipboard/CopyToClipboard.tsx [hunk 3]
```diff
@@ -7,75 +7,52 @@ export enum CopyToClipboardStatus {
...
+    const content = React.useMemo(() => children(status), [children, status]);
+    const timerIdRef = React.useRef<number>();
+    const handleCopy = React.useCallback<Required<ReactCopyToClipboard.Props>['onCopy']>(
+        (copyText, result) => {
+            setStatus(result ? CopyToClipboardStatus.Success : CopyToClipboardStatus.Error);
+
+            if (timerIdRef.current) {
+                clearTimeout(timerIdRef.current);
+            }
+            timerIdRef.current = window.setTimeout(() => {
```
Comment (ogonkov) [inline src/components/CopyToClipboard/CopyToClipboard.tsx:29]: I guess it is would be more clear to have single effect on status change

```jsx
const onCopy = useCallback(() => {setStatus(OTHER_STATUS)}, [])

useEffect(() => {
  if (status === INITIAL_STATUS) {
    return;
  }

  const timer = setTimeout(() => {
    setStatus(INITIAL_STATUS);
  }, timeout);

  return () => {
    clearTimeout(timer);
  };
}, [status]);
```

File: src/components/CopyToClipboard/CopyToClipboard.tsx [hunk 4]
```diff
@@ -7,75 +7,52 @@ export enum CopyToClipboardStatus {
...
-        return (
-            <ReactCopyToClipboard text={String(text)} onCopy={this.handleCopy}>
-                {content}
-            </ReactCopyToClipboard>
-        );
-    }
-
-    componentWillUnmount() {
-        clearTimeout(this.timerId);
+export function CopyToClipboard({children, text, timeout = 1000, onCopy}: CopyToClipboardProps) {
```
Comment (ogonkov) [inline src/components/CopyToClipboard/CopyToClipboard.tsx:17]: Can we add `displayName` please?

File: src/components/CopyToClipboard/CopyToClipboard.tsx (hunk 3 above)
Comment (atroynikov) [reply src/components/CopyToClipboard/CopyToClipboard.tsx]: Moved this part of code to useEffect with status in deps

File: src/components/CopyToClipboard/CopyToClipboard.tsx [hunk 5]
```diff
@@ -7,75 +7,46 @@ export enum CopyToClipboardStatus {
...
+    const content = React.useMemo(() => children(status), [children, status]);
+    const handleCopy = React.useCallback<Required<ReactCopyToClipboard.Props>['onCopy']>(
+        (copyText, result) => {
+            setStatus(result ? CopyToClipboardStatus.Success : CopyToClipboardStatus.Error);
+            onCopy?.(copyText, result);
+        },
+        [onCopy],
+    );
+
+    if (!React.isValidElement(content)) {
```
Comment (ogonkov) [inline src/components/CopyToClipboard/CopyToClipboard.tsx:28]: Not sure that hook could be after exception

File: src/components/CopyToClipboard/CopyToClipboard.tsx [hunk 6]
```diff
@@ -7,75 +7,46 @@ export enum CopyToClipboardStatus {
...
+export function CopyToClipboard({children, text, timeout = 1000, onCopy}: CopyToClipboardProps) {
+    const [status, setStatus] = React.useState(CopyToClipboard.INITIAL_STATUS);
+    const content = React.useMemo(() => children(status), [children, status]);
+    const handleCopy = React.useCallback<Required<ReactCopyToClipboard.Props>['onCopy']>(
+        (copyText, result) => {
+            setStatus(result ? CopyToClipboardStatus.Success : CopyToClipboardStatus.Error);
+            onCopy?.(copyText, result);
+        },
+        [onCopy],
+    );
```
Comment (ValeraS) [inline src/components/CopyToClipboard/CopyToClipboard.tsx:26]: Let's not change the current behaviour. 

```suggestion
    const timer = React.useRef<number>();
    const handleCopy = React.useCallback<Required<ReactCopyToClipboard.Props>['onCopy']>(
        (copyText, result) => {
            setStatus(result ? CopyToClipboardStatus.Success : CopyToClipboardStatus.Error);
            window.clearTimeout(timer.current);
            timer.current = window.setTimeout(() => {
                setStatus(CopyToClipboard.INITIAL_STATUS);
            }, timeout);
            onCopy?.(copyText, result);
        },
        [onCopy, timeout],
    );
    
    React.useEffect(() => () => {
        clearTimeout(timer.current);
    }, []);
```

**Review by ValeraS — CHANGES_REQUESTED** [review]

File: src/components/CopyToClipboard/CopyToClipboard.tsx (hunk 6 above)
Comment (ValeraS) [reply src/components/CopyToClipboard/CopyToClipboard.tsx:26]: Why do it in useEffect when it can be done explicitly in the event handler?

Comment (ogonkov) [reply src/components/CopyToClipboard/CopyToClipboard.tsx:26]: You are using effect anyway to cleanup timer. Keep logic for reseting status in single place looks tidy.

**Comment by amje** [comment]
@atroynikov Is this PR still actual?

**Pull request closed without merging**